syntax = "proto3";

service Raft {
  rpc RequestVote(VoteRequest) returns (RequestVoteResponse) {}
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse) {}
}

message VoteRequest {
  int32 term = 1;
  int32 requester_id = 2;
  int32 last_log_index = 3;
  int32 last_log_term = 4;
}

message RequestVoteResponse {
  int32 term = 1;
  bool vote_granted = 2;
}

message LogEntry {
  int32 term = 1;
  int32 index = 2;
  string key = 3;
  string value = 4;
}

message AppendEntriesRequest {
  int32 term = 1;
  int32 requester_id = 2;
  int32 prev_log_index = 3;
  int32 prev_log_term = 4;
  int32 leader_commit = 5;
  repeated LogEntry log_entries = 6;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
}

service Cluster {
  rpc ProposeMapping(Mapping) returns (ProposeMappingResponse) {}
  rpc GetValue(Key) returns (GetValueResponse) {}
}

message Mapping {
  string key = 1;
  string value = 2;
}

message ProposeMappingResponse {
  bool success = 1;
  optional int32 current_leader = 2;
}

message Key {
  string key = 1;
}

message GetValueResponse {
  string key = 1;
  optional string value = 2;
}
